server/channels/app/import_functions.go

```diff
@@ -1296,58 +1348,80 @@ func (a *App) importAttachment(rctx request.CTX, data *imports.AttachmentImportD
 		}
 		for _, oldFile := range oldFiles {
-			if oldFile.Name != path.Base(name) || oldFile.Size != int64(len(fileData)) {
+			if oldFile.Name != path.Base(name) || oldFile.Size != fileSize {
 				continue
 			}
 
-			// check sha1
-			newHash := sha1.Sum(fileData)
-			oldFileData, err := a.getFileIgnoreCloudLimit(rctx, oldFile.Id)
-			if err != nil {
+			oldFileReader, appErr := a.FileReader(oldFile.Path)
+			if appErr != nil {
 				return nil, model.NewAppError("BulkImport", "app.import.attachment.file_upload.error", map[string]any{"FilePath": *data.Path}, "", http.StatusBadRequest)
 			}
-			oldHash := sha1.Sum(oldFileData)
+			defer oldFileReader.Close()
 
-			if bytes.Equal(oldHash[:], newHash[:]) {
-				rctx.Logger().Info("Skipping uploading of file because name already exists", mlog.String("file_name", name))
+			if ok, err := compareFilesContent(oldFileReader, file, 0); err != nil {
+				rctx.Logger().Error("Failed to compare files content", mlog.String("file_name", name), mlog.Err(err))
+			} else if ok {
+				rctx.Logger().Info("Skipping uploading of file because name already exists and content matches", mlog.String("file_name", name))
 				return oldFile, nil
 			}
+
+			rctx.Logger().Info("File contents don't match, will re-upload", mlog.String("file_name", name))
+
+			// Since compareFilesContent needs to read the whole file we need to
+			// either seek back (local file) or re-open it (zip file).
+			if f, ok := file.(*os.File); ok {
+				rctx.Logger().Info("File is *os.File, can seek")
+				if _, err := f.Seek(0, io.SeekStart); err != nil {
+					return nil, model.NewAppError("BulkImport", "app.import.attachment.seek_file.error", map[string]any{"FilePath": *data.Path}, "", http.StatusBadRequest).Wrap(err)
+				}
+			} else if data.Data != nil {
+				rctx.Logger().Info("File is from ZIP, can't seek, opening again")
+				file.Close()
+
+				f, err := data.Data.Open()
+				if err != nil {
+					return nil, model.NewAppError("BulkImport", "app.import.attachment.bad_file.error", map[string]any{"FilePath": *data.Path}, "", http.StatusBadRequest).Wrap(err)
+				}
+				defer f.Close()
+
+				file = f
+			}
+
+			break
 		}
 	}
 
 	rctx.Logger().Info("Uploading file with name", mlog.String("file_name", name))
 
-	fileInfo, appErr := a.DoUploadFile(rctx, timestamp, teamID, post.ChannelId, post.UserId, name, fileData)
+	fileInfo, appErr := a.UploadFileX(rctx, post.ChannelId, name, file,
+		UploadFileSetTeamId(teamID),
+		UploadFileSetUserId(post.UserId),
+		UploadFileSetTimestamp(timestamp),
+		UploadFileSetContentLength(fileSize),
+	)
 	if appErr != nil {
 		rctx.Logger().Error("Failed to upload file", mlog.Err(appErr), mlog.String("file_name", name))
 		return nil, appErr
 	}
 
-	if fileInfo.IsImage() && !fileInfo.IsSvg() {
-		a.HandleImages(rctx, []string{fileInfo.PreviewPath}, []string{fileInfo.ThumbnailPath}, [][]byte{fileData})
-	}
-
```

Comment:
Why is this not needed?

Reply:
Because it's already part of the upload logic:

https://github.com/mattermost/mattermost/blob/4fd3d6a0e07ce624291b67353c601135cfba095d/server/channels/app/file.go#L787-L794